File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.java
import java.util.*;

class Solution {
    static class Job {
        int req, dur, id;
        Job(int req, int dur, int id) {
            this.req = req;
            this.dur = dur;
            this.id = id;
        }
    }

    public int solution(int[][] jobs) {
        int n = jobs.length;
        if (n == 0) return 0;

        Job[] arr = new Job[n];
        for (int i = 0; i < n; i++) {
            arr[i] = new Job(jobs[i][0], jobs[i][1], i); // req, dur, id
        }

        // 요청시각 기준 정렬
        Arrays.sort(arr, (a, b) -> {
            if (a.req != b.req) return Integer.compare(a.req, b.req);
            if (a.dur != b.dur) return Integer.compare(a.dur, b.dur);
            return Integer.compare(a.id, b.id);
        });

        // 우선순위: dur ↑, req ↑, id ↑
        PriorityQueue<Job> pq = new PriorityQueue<>((a, b) -> {
            if (a.dur != b.dur) return Integer.compare(a.dur, b.dur);
            if (a.req != b.req) return Integer.compare(a.req, b.req);
            return Integer.compare(a.id, b.id);
        });

        long time = 0;
        long total = 0;
        int idx = 0, done = 0;

        while (done < n) {
            while (idx < n && arr[idx].req <= time) {
                pq.offer(arr[idx]);
                idx++;
            }

            if (pq.isEmpty()) {
                // 다음 요청 시각으로 점프
                time = Math.max(time, arr[idx].req);
                continue;
            }

            Job cur = pq.poll();
            time += cur.dur;
            total += (time - cur.req);
            done++;
        }

        return (int)(total / n);
    }
}

File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.kt
import java.util.PriorityQueue
import java.util.Arrays

class Solution {
    fun solution(jobs: Array<IntArray>): Int {
        val n = jobs.size
        if (n == 0) return 0

        // (req, dur, id)
        val arr = Array(n) { i -> intArrayOf(jobs[i][0], jobs[i][1], i) }
        Arrays.sort(arr) { a, b ->
            if (a[0] != b[0]) a[0] - b[0]       // req
            else if (a[1] != b[1]) a[1] - b[1]  // dur
            else a[2] - b[2]                    // id
        }

        // 우선순위: dur ↑, req ↑, id ↑
        val pq = PriorityQueue<IntArray> { a, b ->
            if (a[1] != b[1]) a[1] - b[1]       // dur
            else if (a[0] != b[0]) a[0] - b[0]  // req
            else a[2] - b[2]                    // id
        }

        var time = 0L
        var idx = 0
        var done = 0
        var total = 0L

        while (done < n) {
            while (idx < n && arr[idx][0].toLong() <= time) {
                pq.add(arr[idx])
                idx++
            }

            if (pq.isEmpty()) {
                time = maxOf(time, arr[idx][0].toLong())
                continue
            }

            val cur = pq.poll()
            val req = cur[0]
            val dur = cur[1]

            time += dur.toLong()
            total += (time - req.toLong())
            done++
        }

        return (total / n).toInt()
    }
}